### PR TITLE
Add both_library() to build both shared and static library

### LINKED ISSUE #484: Support building both shared and static libraries at the same time
The `default_library` option should also accept `both` as a valid option which would cause all `library()` invocations to emit build commands to build both static and shared libraries in the same build.

Autotools already does this when you pass `--enable-shared --enable-static` to `configure`.

### PR DESCRIPTION
**Pull request by xclaesse**
Also support default_library='both' to make library() build both shared
and static.

Closes #484

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -594,6 +594,27 @@ class SharedLibraryHolder(BuildTargetHolder):
...
+        self.static_holder = static_holder
+        self.methods.update({'get_shared_lib': self.get_shared_lib_method,
+                             'get_static_lib': self.get_static_lib_method,
+                             })
+
+    def __repr__(self):
+        r = '<{} ({}, {}): {}>'
+        h1 = self.shared_holder.held_object
+        h2 = self.static_holder.held_object
+        return r.format(self.__class__.__name__, h1.get_id(), h2.get_id(), h.filename)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__

undefined name 'h'

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -2269,14 +2293,26 @@ def func_static_lib(self, node, args, kwargs):
...
         return self.build_target(node, args, kwargs, SharedLibraryHolder)
 
+    @permittedKwargs(permitted_kwargs['both_library'])
+    def func_both_lib(self, node, args, kwargs):
+        shared_holder = self.build_target(node, args, kwargs, SharedLibraryHolder)
+
+        kwargs['objects'] = shared_holder.held_object.extract_all_objects()
+        static_holder = self.build_target(node, args, kwargs, StaticLibraryHolder)
+
+        return BothLibraryHolder(shared_holder, static_holder)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__

trailing whitespace

**Comment by xclaesse** [comment]
There is one suboptimal case, the object returned value of library() has different API if built with default_library='both'. Not sure how to make this nicer.

**Comment by xclaesse** [comment]
> There is one suboptimal case, the object returned value of library() has different API if built with default_library='both'. Not sure how to make this nicer.

Fixed that by making BothLibraryHolder subclass BuildTargetHolder and passing it the shared lib target. That means passing it to 'link_with' will link with the shared library.

**Comment by xclaesse** [comment]
Hmm, testing this on glib, I changed all shared_library() to both_library() and libgio fails with this message:

ninja: error: 'gio/gio-2.0@sha/meson-generated_.._gnetworking.h.o', needed by 'gio/libgio-2.0.a', missing and no known rule to make it

everything works if I keep shared_library() just for libgio.

**Comment by nirbheek** [comment]
I like the way this has been implemented, it matches what I imagined this feature would be like and I was going to work in it myself, so thanks and good job!

There's just one hitch: the return value of `library()`. Right now you are defaulting to the shared library for the object to `link_with:`, and people don't always want that.

It covers the most common use-case to link to the shared library and build the static library only for installation, or to link to a few internal utils. However, it is not uncommon to want to link the entire project (and subprojects) statically so your tools are standalone, while also providing shared libraries for installation (I've seen this done in embedded projects, in projects that are used in the initrd, and on Windows).

I think instead of having `both` as the `default-library` option, we should have `shared-static` and `static-shared`. Then, the `held_object` on `BothLibraryHolder` can be changed depending on which one is seleted. Eventually, this explicitness will also help with building Rust targets which have even more types.

**Comment by nirbheek** [comment]
> ninja: error: 'gio/gio-2.0@sha/meson-generated_.._gnetworking.h.o', needed by 'gio/libgio-2.0.a', missing and no known rule to make it

I've seen this bug before elsewhere. It's because when figuring out the list of sources to use in `extract_all_objects()`, we don't skip over headers. See the sources classification in `backends/ninjabackend.py:NinjaBackend.generate_target()`. Specifically, look at how we handle the contents of`target_sources` and `generated_sources`.

This file classification needs to be factored out and probably moved to the `BuildTarget` object so it can also be used by `extract_all_objects()`.

**Comment by xclaesse** [comment]
> I think instead of having both as the default-library option, we should have shared-static and static-shared. Then, the held_object on BothLibraryHolder can be changed depending on which one is selected. Eventually, this explicitness will also help with building Rust targets which have even more types.

I would rather call them both-shared and both-static? Naming stuff is harder than writing code...

**Comment by nirbheek** [comment]
[bikeshed mode on]

> I would rather call them both-shared and both-static? Naming stuff is harder than writing code...

That makes me think "this option says 'both [are] shared'?" What I was thinking with Rust was you'd have `shared-static-rlib` or `rlib-shared-static` or `rlib-static` and so on, so it's more flexible that way to building more than two libraries at once.

**Comment by jibsen** [comment]
> However, it is not uncommon to want to link the entire project (and subprojects) statically so your tools are standalone, while also providing shared libraries for installation (I've seen this done in embedded projects, in projects that are used in the initrd, and on Windows).

I've got at least one project like this, where I build a statically linked executable and a dll.

One thing I'd like to note is that since `default_library` is global, this means if you have one subproject you'd like to build as `both`, you'll have to build your entire project as `both`.

**Comment by nirbheek** [comment]
> One thing I'd like to note is that since default_library is global, this means if you have one subproject you'd like to build as both, you'll have to build your entire project as both.

You can set it per-subproject by using the `default_options:` kwarg to `subproject()` and `dependency()`.

**Comment by jibsen** [comment]
> You can set it per-subproject by using the default_options: kwarg to subproject() and dependency().

That was what I initially thought, but apparently it can only override the default values in the subprojects `meson_options.txt` file. See #2612 (and #2619 which updates the docs).

**Comment by nirbheek** [comment]
Hmm, I don't like how we're handling this `default_options` stuff. It violates the principle of least-surprise and we constantly get issues and questions about it.

**Comment by xclaesse** [comment]
I updated my patch to include doc changes, and use the shared-static/static-shared default_library.

But now that I think about it, I'm wondering if it should be a separate option:
default_library=['shared', 'static', 'both']
default_both_library=['shared', 'static']

**Comment by xclaesse** [comment]
In this last version of the patch, I splitted the setting, to have "both_library" setting that says if it should link static or shared by default. I don't really like that setting name, suggestions?

Also it should fix a bug that it was actually building everything twice because I didn't reset the sources arg when doing the static build. I think that's what was failing CI.

**Comment by jpakkane** [comment]
Maybe we could leave out the possibility of building both types with `library`? Having the return type change based on options is a bit spooky-action-at-a-distancy.

**Comment by jpakkane** [comment]
Thinking about it further, maybe both library holder could have a method `get_default_library_type` or somesuch that would basically reduce to something like this:

    get_option('default_library') == static ? b.get_static_library() : b.get_shared_library()

**Comment by xclaesse** [comment]
> Maybe we could leave out the possibility of building both types with library? Having the return type change based on options is a bit spooky-action-at-a-distancy.

Having a different meaning depending on an option is the whole point of library(). Also the return type is not really different, it can be used exactly the same way than shared/static library. The extra get_static/shared_lib() methods should only be used when using both_library() explicitly.

> Thinking about it further, maybe both library holder could have a method get_default_library_type or somesuch that would basically reduce to something like this:
>
>get_option('default_library') == static ? b.get_static_library() : b.get_shared_library()

I'm not sure when that will be useful, since it already can be used as-is in link_with and it will be using the default. But it can easily be added indeed. `get_default_lib()` to match the naming of the 2 other methods.

**Comment by nirbheek** [comment]
> Thinking about it further, maybe both library holder could have a method get_default_library_type or somesuch that would basically reduce to something like this

Is this because you want to allow people to be explicit about what they are doing or something else? Because the `held_object` in BothLibraryHolder is the default library type.

**Comment by jibsen** [comment]
Now that we have array options, what if `default_library` were an array option instead of a combo, and you could do `default_library : ['static', 'shared']` (and the library linked to by default is the one listed first)?

**Comment by jibsen** [comment]
On second thought, that would probably result in issues with checking the value since there is no way to tell if it is a string or an array.

**Comment by xclaesse** [comment]
Hmmm, CI is still failing and I have no clue why. Any MSVC expert here?

**Comment by nirbheek** [comment]
CI is failing because you're not using `__declspec(dllexport)` on the function prototypes while building. Because of that, `link.exe` does not generate an import library at all since there are no exported symbols.

I'll add a commit to fix it.

**Comment by nirbheek** [comment]
So, while adding that I realized that we need a way to specify in a `declare_dependency()` what flags to use while linking to a library statically and what flags to use when linking dynamically. This is also needed for the current `library()` implementation that can be either static or shared.

The reason is that on Windows, for importing functions from DLLs you only need to specify `__declspec(dllexport)` while building, but for importing variables (constants), you always need to specify `__declspec(dllimport)` otherwise you will get a pointer to the variable instead of the variable itself. This importing is optional for functions.

This means that headers of such libraries support both dynamic and static linking by adding a define like `GST_STATIC_COMPILATION`, etc, that you must set while compiling with the headers used by that library.

We need something like:

```meson
libfoo_dep = declare_dependency(link_with: bothlib,
                                static_c_args: ['-DFOO_STATIC_COMPILATION'],
                                c_args: ['other', 'flags', 'here'])
```

The static args would be added after `c_args` so that you can do the reverse; f.ex. `-DDLL_EXPORT` in `c_args:` and `-UDLL_EXPORT` in `static_c_args:`.

**Comment by nirbheek** [comment]
Note: I do not think that this feature blocks this PR. It's an independent feature, but something to keep in mind.

**Comment by nirbheek** [comment]
What changes do you mean? Github is not showing me the context.

**Comment by xclaesse** [comment]
your latest patch: "tests/common/173: Import constants from a library"

**Comment by nirbheek** [comment]
Well, it's a good example for how to handle that case with MSVC when using `BothLibrary`, which is why I added it.

**Comment by xclaesse** [comment]
Cannot hurt anyway :-)

Would be nice to have that static_c_args trick indeed. I think we should have shared_c_args too fo symetry. So args passed will be shared_c_args+c_args or static_c_args+c_args depending on the type of the link_with. That makes sense to me, and indeed it's not really something new since library() can already return either shared or static and you could want different C args depending on which one is being built.

**Comment by nirbheek** [comment]
I thought about having both `static_c_args` and `shared_c_args`, but decided to keep it simple because every most of the time we try to anticipate what people want, we get it wrong and mess up the API. 

Turns out it's better to implement the minimum needed and grow organically; just have to make sure that we don't restrict what's possible in the future with what we do now.

**Comment by xclaesse** [comment]
Sure, we can always add later when needed, but we cannot remove once it's there.

Just have to make clear in the doc if static_c_args replaces c_args for static builds, or if it's addition to c_args. I think it's the latter, right? Are you working on such a patch already?

**Comment by nirbheek** [comment]
I've learned that it's best to ask for @jpakkane's opinion on something before implementing it, lest the work go to /dev/null. So I'll wait till the feature is absolutely needed, by gstreamer probably.

**Comment by xclaesse** [comment]
@nirbheek Hmm, I rebased my branch and now your commits are gone. I don't know how I can pull them back. Do you still have them ?

**Comment by xclaesse** [comment]
There is something weird when testing this on glib. In glib master we just changed all `shared_library()` to `library()` so we can make static builds.

If I compile with this command, it works fine:
`meson --default-library=static builddir && ninja -C builddir`

This one works fine as well and produce both .a and .so files and .c files are built only once:
`meson --default-library=both builddir && ninja -C builddir`

I can even install then build a static app against libgio with that command:
`gcc test.c -static $(pkg-config gio-2.0 --cflags --libs --static)`

```
$ ldd ./a.out 
	not a dynamic executable
```
But this one fails to link glib's executables:
`meson --default-library=both --default-link=static builddir && ninja -C builddir`
[meson-both-static.txt](https://github.com/mesonbuild/meson/files/1607163/meson-both-static.txt)

**Comment by xclaesse** [comment]
OOh, I think those symbols comes from generated sources. Could be related to PR #2717.

**Comment by xclaesse** [comment]
ping?

**Comment by jpakkane** [comment]
Using object files from one target on the other is a good optimization but unfortunately there is a catch (as there always is). That means the static library will get built with `-fPIC` even if `b_staticpic` is `false`.

**Comment by xclaesse** [comment]
b_staticpic is True by default, and it's up to the user to accept the catch. But I agree using default_library=both and b_staticpic=false should be incompatible. I guess func_both_lib() could raise an exception (or warn) if b_staticpic=false?

**Comment by jpakkane** [comment]
Or detect the case and then just build both libraries from source?

**Comment by jpakkane** [comment]
Looks good, but it really should be `both_libraries` and not `both_library`.

**Comment by xclaesse** [comment]
I called it both_library() to symetry with shared_library(), static_library() and library(). But indeed that could be a plurial. Added a commit to rename it. Thanks.

**Comment by tp-m** [comment]
Would `shared_and_static_library()` be too verbose?

**Comment by xclaesse** [comment]
Grammatically that would be a plural too. I don't think that function will be used much anyway, the main use-case is using library() and with --default-library=both argument. Personally I'm fine with both_libraries().

**Comment by tp-m** [comment]
I wasn't really commenting on the singular/plural aspect, `shared_and_static` just seemed more concise and more discoverable to me than `both` and I hadn't seen it suggested before yet :)

**Comment by QuLogic** [comment]
Could do `dual_library`; not convinced that form needs to be plural, but also not sure that's the most descriptive either.

**Comment by jibsen** [comment]
How do modules fit into this?

**Comment by xclaesse** [comment]
The reason we have library() is to be able to switch all your libraries between shared or static build, but I don't think it makes sense to switch all your libraries into modules. 

But it makes sense to build both a module and static library. For that case I planned to add an extra PR to add a module:true argument to library() which would be the same as using shared_module(), except that if setting default_library=both then it will build a module and static lib.

**Comment by jpakkane** [comment]
We should aim to minimize the number of new options whenever possible since having a gazillion of them is poor usability. Is the default link one really necessary? Could we not make it always match default library? When is the case where different values would be necessary?

**Comment by xclaesse** [comment]
Most of the time you want default-library=both and default-link=shared. Also note that I'm using default-link in PR https://github.com/mesonbuild/meson/pull/2816. The general idea is when we have both shared and static libs available, meson needs to make a decision which one to use.

**Comment by xclaesse** [comment]
After discussion on IRC, I decided to drop `default-link` commit because there is no consensus yet on how to select between shared and static libraries. For now both_libraries() will always return a buildtarget that represents the shared library. Users can manually choose which one to link with code like:

```meson
lib = library()
if get_option('default_library') == 'both'
  lib_dep = lib.get_static_lib()
else
  lib_dep = lib
endif

app = executable(..., dependencies : lib_dep)
```

**Comment by jon-turney** [comment]
The `Cannot extract objects from a target that itself has extracted objects` error this PR adds is reported when trying to build [dpdk](http://dpdk.org/browse/dpdk) (See https://travis-ci.org/jon-turney/meson-corpus-test/jobs/366177420)

@xclaesse I assume that's a genuine instance of this error, but you might want to confirm that.

**Comment by xclaesse** [comment]
Thanks, it's already reported in issue #3401.

**Pull request merged**